lint/clang_tidy.bzl

```diff
@@ -0,0 +1,337 @@
+        exit_code: output file containing the exit code of clang-tidy.
+            If None, then fail the build when clang-tidy exits non-zero.
+    """
+
+    outputs = [stdout]
+    env = {}
+    env["CLANG_TIDY__STDOUT_STDERR_OUTPUT_FILE"] = stdout.path
+    #env["CLANG_TIDY__VERBOSE"] = "1"
+    if exit_code:
+        env["CLANG_TIDY__EXIT_CODE_OUTPUT_FILE"] = exit_code.path
+        outputs.append(exit_code)
+
+    ctx.actions.run_shell(
+        inputs = _gather_inputs(ctx, compilation_context, srcs),
+        outputs = outputs,
+        tools = [executable._clang_tidy_wrapper, executable._clang_tidy],
+        command = executable._clang_tidy_wrapper.path + " $@",
+        arguments = [executable._clang_tidy.path] + _get_args(ctx, compilation_context, srcs),
+        use_default_shell_env = True,
+        env = env,
+        mnemonic = _MNEMONIC,
+        progress_message = "Linting %{label} with clang-tidy",
+    )
+
+def clang_tidy_fix(ctx, compilation_context, executable, srcs, patch, stdout, exit_code):
+    """Create a Bazel Action that spawns clang-tidy with --fix.
+
+    Args:
+        ctx: an action context OR aspect context
+        compilation_context: from target
+        executable: struct with a clang_tidy field
+        srcs: list of file objects to lint
+        patch: output file containing the applied fixes that can be applied with the patch(1) command.
+        stdout: output file containing the stdout or --output-file of clang-tidy
+        exit_code: output file containing the exit code of clang-tidy
+    """
+    patch_cfg = ctx.actions.declare_file("_{}.patch_cfg".format(ctx.label.name))
+
+    args = _get_args(ctx, compilation_context, srcs)
+    ctx.actions.write(
+        output = patch_cfg,
+        content = json.encode({
+            "linter": executable._clang_tidy_wrapper.path,
+            "args": [executable._clang_tidy.path, "--fix"] + _get_args(ctx, compilation_context, srcs),
+            "env": {
+                #"CLANG_TIDY__VERBOSE": "1",
+            },
+            "files_to_diff": [src.path for src in srcs],
+            "output": patch.path,
+        }),
+    )
+
+    ctx.actions.run(
+        inputs = _gather_inputs(ctx, compilation_context, srcs) + [patch_cfg],
+        outputs = [patch, stdout, exit_code],
+        executable = executable._patcher,
+        arguments = [patch_cfg.path],
+        env = {
+            "BAZEL_BINDIR": ".",
+            "JS_BINARY__EXIT_CODE_OUTPUT_FILE": exit_code.path,
+            "JS_BINARY__STDOUT_OUTPUT_FILE": stdout.path,
+            "JS_BINARY__SILENT_ON_SUCCESS": "1",
+        },
+        tools = [executable._clang_tidy_wrapper, executable._clang_tidy],
+        mnemonic = _MNEMONIC,
+        progress_message = "Linting %{label} with clang-tidy",
+    )
+
+# buildifier: disable=function-docstring
+def _clang_tidy_aspect_impl(target, ctx):
+    if not CcInfo in target:
+        return []
+
+    # todo: keep this? #Ignore external targets
```

Comment:
we certainly shouldn't produce lint results for source files that the user doesn't control. Do we need to uncomment this? Or do you need to use a `filter_srcs` helper like the sibling linters?

Reply:
Removed commented blocks, using filter_srcs